src/track/beats.cpp

```diff
@@ -1,29 +1,538 @@
+    return m_beatsInternal.findClosestBeat(frame);
+}
+
+Beat Beats::findNthBeat(FramePos frame, int n) const {
+    QMutexLocker locker(&m_mutex);
+    return m_beatsInternal.findNthBeat(frame, n);
+}
+
+std::unique_ptr<Beats::iterator> Beats::findBeats(
+        FramePos startFrame, FramePos stopFrame) const {
+    QMutexLocker locker(&m_mutex);
+    return m_beatsInternal.findBeats(startFrame, stopFrame);
+}
+
+Bpm Beats::getBpm() const {
+    QMutexLocker locker(&m_mutex);
+    return m_beatsInternal.getBpm();
+}
+
+Bpm Beats::getBpmAroundPosition(FramePos curFrame, int n) const {
+    QMutexLocker locker(&m_mutex);
+    return m_beatsInternal.getBpmAroundPosition(curFrame, n);
+}
+
+void Beats::setSignature(TimeSignature sig, int beatIndex) {
+    QMutexLocker locker(&m_mutex);
+    m_beatsInternal.setSignature(sig, beatIndex);
+    locker.unlock();
+    emit updated();
+}
+
+void Beats::translate(FrameDiff_t numFrames) {
+    QMutexLocker locker(&m_mutex);
+    m_beatsInternal.translate(numFrames);
+    locker.unlock();
+    emit updated();
+}
+
+void Beats::scale(enum BeatsInternal::BPMScale scale) {
+    QMutexLocker locker(&m_mutex);
+    m_beatsInternal.scale(scale);
+    locker.unlock();
+    emit updated();
+}
+
+void Beats::setBpm(Bpm bpm, int beatIndex) {
+    QMutexLocker locker(&m_mutex);
+    m_beatsInternal.setBpm(bpm, beatIndex);
+    locker.unlock();
+    emit updated();
+}
+
+FramePos Beats::getFirstBeatPosition() const {
+    return m_beatsInternal.getFirstBeatPosition();
+}
+
+FramePos Beats::getLastBeatPosition() const {
+    return m_beatsInternal.getLastBeatPosition();
+}
+
+SINT Beats::getSampleRate() const {
+    return m_track->getSampleRate();
+}
+
+void Beats::setAsDownbeat(int beatIndex) {
+    QMutexLocker locker(&m_mutex);
+    m_beatsInternal.setAsDownbeat(beatIndex);
+    locker.unlock();
+    emit updated();
+}
+
+QDebug operator<<(QDebug dbg, const BeatsPointer& arg) {
+    dbg << arg->m_beatsInternal;
+    return dbg;
+}
+
+QDebug operator<<(QDebug dbg, const BeatsInternal& arg) {
+    QVector<FramePos> beatFramePositions;
+    for (const auto& beat : arg.m_beats) {
+        beatFramePositions.append(beat.getFramePosition());
+    }
+    dbg << "["
+        << "Aggregate BPM:" << arg.m_bpm << "|"
+        << "Number of beats:" << arg.m_beats.size() << "|"
+        << "Beats:" << beatFramePositions << "]";
+    return dbg;
+}
+Beat BeatsInternal::findNthBeat(FramePos frame, int n) const {
+    if (!isValid() || n == 0) {
+        return kInvalidBeat;
+    }
+
+    Beat beat(frame);
+    // it points at the first occurrence of beat or the next largest beat
+    BeatList::const_iterator it =
+            std::lower_bound(m_beats.cbegin(), m_beats.cend(), beat);
+
+    // If the position is within 1/10th of the average beat length,
+    // pretend we are on that beat.
+    const double kFrameEpsilon =
+            kBeatVicinityFactor * getBeatLengthFrames(getBpm(), m_iSampleRate);
```

Comment:
Using `getBpm` here assumes the track is one constant tempo. The local BPM should be used.